File: backend/api/postulaciones.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from backend.infrastructure.persistence.database import get_db
from backend.domain.services.postulacion_service import PostulacionService
from backend.infrastructure.persistence.repositories.postulacion_repo import PostulacionRepository
from backend.schemas.postulacion import PostulacionCreate, PostulacionUpdateEstado, PostulacionResponse
from backend.schemas.traza import TrazaResponse

router = APIRouter()


def _service(db: Session = Depends(get_db)) -> PostulacionService:
    return PostulacionService(db)


@router.post("/", response_model=PostulacionResponse, status_code=201)
def crear(data: PostulacionCreate, svc: PostulacionService = Depends(_service)):
    try:
        return svc.crear_postulacion(data)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))


@router.get("/", response_model=list[PostulacionResponse])
def listar_todas(svc: PostulacionService = Depends(_service)):
    """Lista todas las postulaciones del sistema. Para uso del admin."""
    return svc.listar_todas()


@router.get("/perfil/{perfil_id}", response_model=list[PostulacionResponse])
def listar_por_perfil(perfil_id: str, svc: PostulacionService = Depends(_service)):
    return svc.listar_por_perfil(perfil_id)


@router.get("/{postulacion_id}", response_model=PostulacionResponse)
def obtener(postulacion_id: str, svc: PostulacionService = Depends(_service)):
    post = svc.obtener(postulacion_id)
    if not post:
        raise HTTPException(status_code=404, detail="Postulación no encontrada")
    return post


@router.patch("/{postulacion_id}/estado", response_model=PostulacionResponse)
def cambiar_estado(
    postulacion_id: str,
    data: PostulacionUpdateEstado,
    svc: PostulacionService = Depends(_service),
):
    post = svc.cambiar_estado(postulacion_id, data.estado, data.notas)
    if not post:
        raise HTTPException(status_code=404, detail="Postulación no encontrada")
    return post


@router.get("/{postulacion_id}/trazas", response_model=list[TrazaResponse])
def listar_trazas(postulacion_id: str, db: Session = Depends(get_db)):
    repo = PostulacionRepository(db)
    return repo.get_trazas(postulacion_id)

File: frontend/src/api/postulaciones.js
import { apiFetch } from "./client"

const BASE = ""

export async function crearPostulacion(data) {
  return apiFetch(`${BASE}/postulaciones/`, {
    method: "POST",
    body: JSON.stringify(data),
  })
}

export async function listarPostulaciones(perfil_id) {
  return apiFetch(`${BASE}/postulaciones/perfil/${perfil_id}`)
}

export async function listarTodasPostulaciones() {
  return apiFetch(`${BASE}/postulaciones/`)
}

export async function obtenerPostulacion(id) {
  return apiFetch(`${BASE}/postulaciones/${id}`)
}

export async function cambiarEstado(id, estado, notas = null) {
  return apiFetch(`${BASE}/postulaciones/${id}/estado`, {
    method: "PATCH",
    body: JSON.stringify({ estado, notas }),
  })
}

export async function obtenerTrazas(perfil_id, limit = 50) {
  return apiFetch(`${BASE}/trazas/${perfil_id}?limit=${limit}`)
}

export async function ejecutarPipeline(perfil_id) {
  return apiFetch(`${BASE}/pipeline/ejecutar`, {
    method: "POST",
    body: JSON.stringify({ perfil_id }),
  })
}